packages/qwik/src/core/util/string.ts

```diff
@@ -0,0 +1,144 @@
+'./string.unit.ts';
+// pack bytes into valid UTF-16 string
+//
+// strategy:
+//
+// * using 0xFFFD as the escape character
+//  * if there is 0xFFFD in the bytes, double it
+// * if there is unmatched surrogate pair, mark it by the escape character
+// * and put a fake surrogate pair to make it valid
+//  * 0xD800 for fake high surrogate to be with unmatched low surrogate
+//  * 0xDC00 for fake low surrogate to be with unmatched high surrogate
+//
+// if the unmatched high surrogate is 0xD800, it is collided with the fake
+// high surrogate, so use [0xD801, 0xDC01] as the fake surrogate pair
+// representing the 0xD800.
+//
+// If the length of the bytes is odd, the last byte is put after the escape
+// character. As the bytes after the escape character are in 0xD800 to 0xDFFF,
+// we can distingwish the last byte by its high byte being 0x00.
+//
+export const packUint8Array = (bytes: Uint8Array) => {
+  const odd = bytes.length % 2 === 1;
+  const dbytes = new Uint16Array(bytes.buffer, 0, bytes.length >> 1);
+  let code = '';
+  let surrogate = false;
+  for (let i = 0; i < dbytes.length; i++) {
+    const c = dbytes[i];
+    // test high surrogate
+    if (c >= 0xd800 && c <= 0xdbff) {
+      if (surrogate) {
+        // unmatched high surrogate
+        const prev = dbytes[i - 1];
+        const [hi, lo] = prev === 0xd800 ? [0xd801, 0xdc01] : [prev, 0xdc00];
+        // put the 0xFFFD and the fake surrogate pair to make it valid
+        code += String.fromCharCode(0xfffd, hi, lo);
+        // keep surrogate is true because c is high surrogate
+      }
+      surrogate = true;
+      continue;
+    }
+    // test low surrogate
+    if (c >= 0xdc00 && c <= 0xdfff) {
+      if (surrogate) {
+        // valid surrogate pair
+        code += String.fromCharCode(dbytes[i - 1], c);
+        surrogate = false;
+        continue;
+      }
+      // unmatched low surrogate
+      // put the 0xFFFD and the fake high surrogate to make it valid
+      code += String.fromCharCode(0xfffd, 0xd800, c);
+      continue;
+    }
+    if (surrogate) {
+      // no low surrogate after high surrogate
+      const prev = dbytes[i - 1];
+      const [hi, lo] = prev === 0xd800 ? [0xd801, 0xdc01] : [prev, 0xdc00];
+      // put the 0xFFFD and the fake surrogate pair to make it valid
+      code += String.fromCharCode(0xfffd, hi, lo);
+      surrogate = false; // reset surrogate
+    }
+    // escape the BOM
+    if (c === 0xfeff) {
+      // BOM
+      code += String.fromCharCode(0xfffd, 0xd801, 0xdc02);
```

Comment:
this could be `FFFD 1000`, but is it even needed? The serialization will never start with the BOM.

Reply:
Sorry, I couldn't understand what could be `FFFD 1000`?
This code is escaping the `0xfe 0xff` in the Uint8Array.
This 2 bytes is same to the BOM character in the UTF-16, so some environments meddlingly delete them.
So we have to save them.

Reply:
I meant that FFFD 1000 is still available for escaping after you handle the unmatched pairs, but if you do it with FFFD F999 or so, so still only escaping is needed for FFFD F...

Reply:
Sorry, as of the change of the algorithm, this code is outdated.